cloudinit/sources/DataSourceVMware.py

```diff
@@ -0,0 +1,780 @@
+        if addr4_fams:
+            af_inet4 = addr4_fams.get(netifaces.AF_INET)
+            if af_inet4:
+                if len(af_inet4) > 1:
+                    LOG.warning(
+                        "device %s has more than one ipv4 address: %s",
+                        dev4,
+                        af_inet4,
+                    )
+                elif "addr" in af_inet4[0]:
+                    ipv4 = af_inet4[0]["addr"]
+
+    # Try to get the default IPv6 address by first seeing if there is a default
+    # IPv6 route.
+    gw6 = default_gw.get(netifaces.AF_INET6)
+    if gw6:
+        _, dev6 = gw6
+        addr6_fams = netifaces.ifaddresses(dev6)
+        if addr6_fams:
+            af_inet6 = addr6_fams.get(netifaces.AF_INET6)
+            if af_inet6:
+                if len(af_inet6) > 1:
+                    LOG.warning(
+                        "device %s has more than one ipv6 address: %s",
+                        dev6,
+                        af_inet6,
+                    )
+                elif "addr" in af_inet6[0]:
+                    ipv6 = af_inet6[0]["addr"]
+
+    # If there is a default IPv4 address but not IPv6, then see if there is a
+    # single IPv6 address associated with the same device associated with the
+    # default IPv4 address.
+    if ipv4 and not ipv6:
+        af_inet6 = addr4_fams.get(netifaces.AF_INET6)
+        if af_inet6:
+            if len(af_inet6) > 1:
+                LOG.warning(
+                    "device %s has more than one ipv6 address: %s",
+                    dev4,
+                    af_inet6,
+                )
+            elif "addr" in af_inet6[0]:
+                ipv6 = af_inet6[0]["addr"]
+
+    # If there is a default IPv6 address but not IPv4, then see if there is a
+    # single IPv4 address associated with the same device associated with the
+    # default IPv6 address.
+    if not ipv4 and ipv6:
+        af_inet4 = addr6_fams.get(netifaces.AF_INET)
+        if af_inet4:
+            if len(af_inet4) > 1:
+                LOG.warning(
+                    "device %s has more than one ipv4 address: %s",
+                    dev6,
+                    af_inet4,
+                )
+            elif "addr" in af_inet4[0]:
+                ipv4 = af_inet4[0]["addr"]
+
+    return ipv4, ipv6
+
+
+# patched socket.getfqdn() - see https://bugs.python.org/issue5004
+
+
+def getfqdn(name=""):
+    """Get fully qualified domain name from name.
+    An empty argument is interpreted as meaning the local host.
+    """
+    name = name.strip()
+    if not name or name == "0.0.0.0":
+        name = socket.gethostname()
+    try:
+        addrs = socket.getaddrinfo(
+            name, None, 0, socket.SOCK_DGRAM, 0, socket.AI_CANONNAME
+        )
+    except socket.error:
+        pass
+    else:
+        for addr in addrs:
+            if addr[3]:
+                name = addr[3]
+                break
+    return name
+
+
+def is_valid_ip_addr(val):
```

Comment:
We are doing a lot of re-inventing, or improving on existing cloud-init utility functions in this datasource. While I appreciate the motivations here, I'm trying to review this section in more detail (as well as general netifaces imports) to see what we can re-use to avoid instrumenting network detection logic that needs to be maintained in two places in cloud-init. 


I need another day to get through this review, so I'll follow up tomorrow.

Reply:
FWIW, I've explained to others that I view CI and its datasources like an NES or SNES. CI is the console, and the datasources are the cartridges. Some datasources, like some cartridges, have extra functionality, i.e. specialized chips, that enhance the overall user experience. With respect to cloud-init, I always assumed this was on purpose.

But, you _are_ right that I've done some things that _could_ be in core CI. However, some of what I've done is somewhat unique to the VMware situation where networking derived from DHCP still should be present in the instance data. That and I felt that netifaces offered a more consistent and deterministic way to inspect the local host's networking stack that shell exec'ing to a series of ip commands.

Reply:
Thanks for this repeated context here @akutz. I see I missed some of your earlier resolved responses on this PR.

@TheRealFalcon and I talked a bit about my minor misgivings about instrumenting network-specific helper functions within VMware datasource.

TLDR: would you be opposed to documenting a couple of TODO(...)  comments in the VMwareDatasource about long-term direction to try aligning with common cloud-init network utilities for the functions:
  - is_valid_ip_addr -> # TODO( extend cloudinit.net.is_ip_addr to ignore link_local/loopback or unspecified to drop this func)
  - 

I realize that my friction here is the following:
 * If we instrument duplicated (but better) detection logic down in VMware datasource, shared CI networking detection functions miss out on any improvements available via `netifaces`. Alternately, VMWare may miss out on corner cases that common CI network functions improve handling of and so we find ourselves trying to maintain two "separate but equal" solutions over time, which rarely happens unless we hit critical bugs that point us to the secondary implementations.
    * This example of dual-maintenance showed up a couple of times in long-term maintenance of the cloudinit/sources/helpers/vmware/imc modules which had a number of VMware-specific helper functions which ultimately do the same thing as cloud-init's common net functions

My misgivings on the network-related functions are more of a goal than a firm blocker to your work and can be deemed as a "nice to have" or TODO for subsequent iterations. I realize that you and other VMware devs will be best placed to iterate and maintain this code long-term given your access to VMware platforms and products far exceeds what upstream cloud-init devs have access to, so upstream committers are ill-placed to validate changes to this datasource functions. As a result I'd leave most of that validation work to whatever integration  testing runs you are able to perform on your side in our subsequent reviews.

That said, I think there are a couple of places that we can use common cloud-init utils if possible, the rest of the networking-related functions let's table for a subsequent iteration and just add some TODO breadcrumbs to get us thinking about that alignment in the future. I'll add those set of suggestions inline over the next hour and probably conslidate a patch with the diff of suggestions for your review.

Reply:
> Thanks for this repeated context here @akutz. I see I missed some of your earlier resolved responses on this P
Not at all. I appreciate your time on this!

> TLDR: would you be opposed to documenting a couple of TODO(...) comments in the VMwareDatasource about long-term direction to try aligning with common cloud-init network utilities
Of course not, happy to do it.

Your concerns are reasonable, and honestly, if you decided to block this DS from merging until we sorted out everything to exactly the way you would like it, I would understand.

My primary goals are:

* Stop maintaining this DS in a separate repo
* Provide backwards compatibility for the user experience, even as we update the internals
* Make it easier for others to have eyes on it

I sense you are also aligned with these goals, so I'm fine with whatever you want to do here. Obviously the sooner it's merged the sooner more people can work to address the issues, but again, that's totally up to y'all. I just appreciate all of your help on this (and @TheRealFalcon 's help as well)!